### PR TITLE
HTTP lookup source max retries

### PR DESCRIPTION
**Pull request by Raz0r**
#### Description

Lookup queries are now retried in case of IOException up to `gid.connector.http.source.lookup.max-retries` with a delay of `gid.connector.http.source.lookup.request.retry-timeout-ms` between retries. The default values are 3 retries and 1 second delay.

Resolves https://github.com/getindata/flink-http-connector/issues/66

##### PR Checklist
- [x] Tests added
- [x] [Changelog](CHANGELOG.md) updated

### CODE REVIEW

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java [hunk 1]
```diff
@@ -27,6 +27,10 @@
 @Slf4j
 public class JavaNetHttpPollingClient implements PollingClient<RowData> {
 
+    public static final String DEFAULT_REQUEST_MAX_RETRIES = "3";
+
+    public static final String DEFAULT_REQUEST_RETRY_TIMEOUT_MS = "1000";
```
Comment (davidradl) [inline src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java:32]: I suggest calling this DEFAULT_REQUEST_RETRY_INTERVAL_MS as it is not a timeout.

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java [hunk 2]
```diff
@@ -27,6 +27,10 @@
 @Slf4j
 public class JavaNetHttpPollingClient implements PollingClient<RowData> {
 
+    public static final String DEFAULT_REQUEST_MAX_RETRIES = "3";
```
Comment (davidradl) [inline src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java:30]: should this not be an int?

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java [hunk 3]
```diff
@@ -74,15 +96,43 @@ public Optional<RowData> pull(RowData lookupRow) {
         }
     }
 
-    // TODO Add Retry Policy And configure TimeOut from properties
```
Comment (davidradl) [inline src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java:77]: I think the intent of this comment is to make use of the Flink Retry strategies, which includes async.

**Review by davidradl — CHANGES_REQUESTED** [review]
I have made some comments - I think we should have consistent retry mechanisms for sink sourcs async synch. There are Retry strategies in Flink e.g. AsyncRetryStrategies.  I would hope we could make an IOException bubble up so that the retry strategies would retry . 

In FlinkI  see
   ```
 @JsonTypeName("RetryLookupOptions")
    public static class RetryLookupOptions {
        public static final String FIELD_NAME_RETRY_PREDICATE = "retry-predicate";
        public static final String FIELD_NAME_RETRY_STRATEGY = "retry-strategy";
        public static final String FIELD_NAME_RETRY_FIXED_DELAY = "fixed-delay";
        public static final String FIELD_NAME_RETRY_MAX_ATTEMPTS = "max-attempts";

 
```

**Comment by Raz0r** [comment]
Hey @davidradl thanks, didn't know about existing `AsyncRetryStrategies`, I will need some time to figure out how to use it correctly with making IOException bubble up.

**Comment by davidradl** [comment]
@Raz0r See [https://github.com/getindata/flink-http-connector/pull/94](https://github.com/getindata/flink-http-connector/pull/94). This gives an example on how to use the Partial caching in this area.

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java (hunk 2 above)
Comment (kristoffSC) [reply src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java:30]: We tend keep all properties as Strings, becase they are passed as Properties<String,String> from Table factories.

Later we use them like this:

```
this.httpRequestMaxRetries = Integer.parseInt(
            options.getProperties().getProperty(
                HttpConnectorConfigConstants.LOOKUP_HTTP_MAX_RETRIES,
                DEFAULT_REQUEST_MAX_RETRIES
            )
        );
```

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java (hunk 3 above)
Comment (kristoffSC) [reply src/main/java/com/getindata/connectors/http/internal/table/lookup/JavaNetHttpPollingClient.java:77]: The thing is that AsyncSink uses totally different, dedicated mechanism for retries => Flink native support. 
And actually there is separate TODO for this in HttpSinkWriter.

What kind of Flink retry strategies you were here @davidradl ?

**Comment by Raz0r** [comment]
Closing this in favor of https://github.com/getindata/flink-http-connector/pull/129 and https://github.com/getindata/flink-http-connector/pull/120

**Pull request closed without merging**